R/calculate_silhouette_width.R

```diff
@@ -44,36 +42,95 @@ calculate_silhouette_width <- function(merged_sce,
     stop("The fraction of cells to downsample should be between 0 and 1.")
   }
 
-  # Pull out the PCs or analogous reduction
-  pcs <- reducedDim(merged_sce, pc_name)
-
   # Check that `batch_column` is in colData of SCE
   if (!batch_column %in% colnames(colData(merged_sce))) {
     stop("The specified batch column is missing from the colData of the SingleCellExperiment object.")
   }
 
-  # Remove batch NAs from PCs and label rownames
-  labeled_pcs <- filter_pcs(pcs, colData(merged_sce)[, batch_column])
-
-  # Perform calculations
-  all_silhouette <- purrr::map(1:nreps, \(rep) {
-    # Downsample PCs
-    downsampled <- downsample_pcs(labeled_pcs, frac_cells)
-    # Calculate batch ASW and add into final tibble
-    bluster::approxSilhouette(downsampled, rownames(downsampled)) |>
-      tibble::as_tibble() |>
-      dplyr::mutate(rep = rep) |>
-      dplyr::select(
-        "rep",
-        silhouette_width = "width",
-        silhouette_cluster = "cluster",
-        other_cluster = "other"
-      )
-  }) |>
-    dplyr::bind_rows() |>
-    # Add integration method into tibble
-    dplyr::mutate(pc_name = pc_name)
+  # Calculate the silhouette width values across list of PCs
+  all_silhouette_df <- purrr::map(
+    pc_names,
+    \(pcs)
+    silhouette_width_from_pcs(
+      merged_sce = merged_sce,
+      batch_column = batch_column,
+      pc_name = pcs,
+      frac_cells = frac_cells,
+      nreps = nreps,
+      seed = seed
+    )
+  ) |>
+    dplyr::bind_rows()
 
   # Return tibble with silhouette width results which can further be summarized downstream
-  return(all_silhouette)
+  return(all_silhouette_df)
 }
+
+
+#' Calculate silhouette width scores using PCs from a merged SCE object for use
+#' in integration metric calculations
+#'
+#' @param merged_sce The merged SCE object containing data from multiple batches
+#' @param pc_name The name used to access the PCA results stored in the
+#'   SingleCellExperiment object
+#' @param batch_column The variable in `merged_sce` indicating the grouping of interest.
+#'  Generally this is either batches or cell types. Default is "library_id".
+#' @param frac_cells The fraction of cells to downsample to.
+#' @param nreps The number of times to repeat sub-sampling procedure.
+#' @param seed Seed for initializing random sampling. Default is NULL.
+#'
+#' @return Tibble with five columns: `rep`, representing the given downsampling replicate;
+#'   `silhouette_width`, the calculated silhouette width for the given `rep`; `silhouette_cluster`,
+#'   the assigned cluster for the cell during silhouette calculation, i.e. the true identity;
+#'   `other_cluster`, the other assigned for the cell during silhouette calculation;
+#'   `pc_name`, the name associated with the pc results
+silhouette_width_from_pcs <-
+  function(merged_sce,
+           pc_name,
+           batch_column = "library_id",
+           frac_cells,
+           nreps,
```

Comment:
Pretty minor, but it is generally good practice to put required parameters first, with the ones that have default values following.

Alternatively, you could leave the order as is, but add defaults to `frac_cells` and `nreps` as above.